compiler/src/main/scala/org/scalajs/nscplugin/PrepJSExports.scala

```diff
@@ -433,55 +442,37 @@ trait PrepJSExports[G <: Global with Singleton] { this: PrepJSInterop[G] =>
     }
   }
 
-  /** generate an exporter for a DefDef including default parameter methods */
-  private def genExportDefs(defSym: Symbol, jsName: String, pos: Position) = {
-    val clsSym = defSym.owner
-    val scalaName =
-      jsInterop.scalaExportName(jsName, jsInterop.isJSProperty(defSym))
+  /** generate an exporter for a target including default parameter methods */
+  private def genExportDefs(sym: Symbol, jsName: String, pos: Position) = {
+    val clsSym =
+      if (sym.isConstructor) sym.owner.owner
+      else sym.owner
+
+    val isProperty = sym.isModuleClass || isJSAny(sym) || jsInterop.isJSProperty(sym)
+    val scalaName = jsInterop.scalaExportName(jsName, isProperty)
 
     // Create symbol for new method
-    val expSym = defSym.cloneSymbol
-
-    // Set position of symbol
-    expSym.pos = pos
-
-    // Alter type for new method (lift return type to Any)
-    // The return type is lifted, in order to avoid bridge
-    // construction and to detect methods whose signature only differs
-    // in the return type.
-    // Attention: This will cause boxes for primitive value types and value
-    // classes. However, since we have restricted the return types, we can
-    // always safely remove these boxes again in the back-end.
-    if (!defSym.isConstructor)
-      expSym.setInfo(retToAny(expSym.tpe))
-
-    // Change name for new method
-    expSym.name = scalaName
-
-    // Update flags
-    expSym.setFlag(Flags.SYNTHETIC)
-    expSym.resetFlag(
-        Flags.DEFERRED     | // We always have a body
-        Flags.ACCESSOR     | // We are never a "direct" accessor
-        Flags.CASEACCESSOR | // And a fortiori not a case accessor
-        Flags.LAZY         | // We are not a lazy val (even if we export one)
-        Flags.OVERRIDE       // Synthetic methods need not bother with this
-    )
-
-    // Remove export annotations
-    expSym.removeAnnotation(JSExportAnnotation)
-
-    // Add symbol to class
-    clsSym.info.decls.enter(expSym)
+    val expSym = clsSym.newMethod(scalaName, pos, Flags.SYNTHETIC)
```

Comment:
I'm pretty sure we should copy the visibility modifiers (flags and/or `privateWithin`) of `sym`. Previously that was done by `defSym.cloneSymbol`.

Reply:
They're public to JS, but not to the Scala code. It may not make a true difference, since users *should* not try to access them. Still, I believe it's more consistent to copy the visibility of the original member, perhaps more because that's what we did before than because of a real rationale.

Perhaps it's somewhat important not to *change* for compatibility purposes. If a superclass `A` in library 1 had a protected override, and subclass `B extends A` in lib 2 overrode and re-exported that method as protected again, the exporter would be protected in `B`. After `A` is recompiled with the new Scala.js version, the exporter becomes public; but then `B` has a protected method that overrides a public method, which is not allowed.

Reply:
Of course you're right. We have a similar problem with `final` modifiers as well: They are not copied anymore, essentially always allowing to override a `JSExport`ed method (I'm not sure it's a problem, but IMO its definitely a noteworthy change).

I need to think about how to address the visibility issue. Maybe just take the modifiers of the class, never the ones of the constructor? Or just always make class exports public?

Reply:
I think taking the modifiers of the class would make the most sense.